src/firefox/rdp-client.js

```diff
@@ -0,0 +1,303 @@
+  resolve: Function,
+  reject: Function,
+|}
+
+type ParseResult = {|
+  data: Buffer,
+  rdpMessage?: Object,
+  error?: Error,
+  fatal?: boolean,
+|}
+
+export const DEFAULT_PORT = 6000;
+export const DEFAULT_HOST = '127.0.0.1';
+
+const UNSOLICITED_EVENTS = new Set([
+  'tabNavigated',
+  'styleApplied',
+  'propertyChange',
+  'networkEventUpdate',
+  'networkEvent',
+  'propertyChange',
+  'newMutations',
+  'frameUpdate',
+  'tabListChanged',
+]);
+
+// Parse RDP packets: BYTE_LENGTH + ':' + DATA.
+export function parseRDPMessage(data: Buffer): ParseResult {
+  const str = data.toString();
+  const sepIdx = str.indexOf(':');
+  if (sepIdx < 1) {
+    return {data};
+  }
+
+  const byteLen = parseInt(str.slice(0, sepIdx));
+  if (isNaN(byteLen)) {
+    const error = new Error('Error parsing RDP message length');
+    return {data, error, fatal: true};
+  }
+
+  if (data.length - (sepIdx + 1) < byteLen) {
+    // Can't parse yet, will retry once more data has been received.
+    return {data};
+  }
+
+  data = data.slice(sepIdx + 1);
+  const msg = data.slice(0, byteLen);
+  data = data.slice(byteLen);
+
+  try {
+    return {data, rdpMessage: JSON.parse(msg.toString())};
+  } catch (error) {
+    return {data, error, fatal: false};
+  }
+}
+
+export function connectToFirefox(port: number): Promise<FirefoxRDPClient> {
+  const client = new FirefoxRDPClient();
+  return client.connect(port).then(() => client);
+}
+
+export default class FirefoxRDPClient extends EventEmitter {
+  _incoming: Buffer;
+  _pending: Array<{| request: RDPRequest, deferred: Deferred |}>;
+  _active: Map<string, Deferred>;
+  _rdpConnection: net.Socket;
+  _onData: Function;
+  _onError: Function;
+  _onEnd: Function;
+  _onTimeout: Function;
+
+  constructor() {
+    super();
+    this._incoming = Buffer.alloc(0);
+    this._pending = [];
+    this._active = new Map();
+
+    this._onData = this.onData.bind(this);
+    this._onError = this.onError.bind(this);
+    this._onEnd = this.onEnd.bind(this);
+    this._onTimeout = this.onTimeout.bind(this);
+  }
+
+  connect(port: number): Promise<void> {
+    return new Promise((resolve, reject) => {
+      // Create a domain to wrap the errors that may be triggered
+      // by creating the client connection (e.g. ECONNREFUSED)
+      // so that we can reject the promise returned instead of
+      // exiting the entire process.
+      const d = domain.create();
+      d.once('error', reject);
+      d.run(() => {
+        const conn = net.createConnection({
+          port,
+          host: DEFAULT_HOST,
+        });
+
+        this._rdpConnection = conn;
+        conn.on('data', this._onData);
+        conn.on('error', this._onError);
+        conn.on('end', this._onEnd);
+        conn.on('timeout', this._onTimeout);
+
+        // Resolve once the expected initial root message
+        // has been received.
+        this._expectReply('root', {resolve, reject});
+      });
+    });
+  }
+
+  disconnect(): Promise<void> {
```

Comment:
Why is `disconnect` async when the implementation is not? There are some callers (in the implementation and in the test) that treat `disconnect` as synchronous, while others do not treat it as async. I suggest to be consistent.

At this point, I guess that it's better to make it synchronous, as there is no wait for the underlying connection to really be closed.

Reply:
Fine by me.

I don't actually recall right now if I did make it async because it was in the origin firefox-client implementation or just because I was planning to make it await until the connection was really closed, anyway I'll just make it sync for now (and we can evaluate to make it async if we decide to let it wait for the actual disconnection)